src/nvim/api/vimscript.c

```diff
@@ -45,13 +45,33 @@
 /// @see |nvim_cmd()|
 ///
 /// @param src      Vimscript code
-/// @param output   Capture and return all (non-error, non-shell |:!|) output
+/// @param opts  Optional parameters.
+///           - output: (boolean, default false) Whether to capture and return
+///                     all (non-error, non-shell |:!|) output.
 /// @param[out] err Error details (Vim error), if any
-/// @return Output (non-error, non-shell |:!|) if `output` is true,
-///         else empty string.
-String nvim_exec(uint64_t channel_id, String src, Boolean output, Error *err)
-  FUNC_API_SINCE(7)
+/// @return Dictionary containing information about execution, with these keys:
+///       - output: (string) Output if `opts.output` is true, empty string
+///                 otherwise.
+Dictionary nvim_exec2(uint64_t channel_id, String src, Dict(exec_opts) *opts, Error *err)
+  FUNC_API_SINCE(11)
 {
+  Dictionary result = ARRAY_DICT_INIT;
+  PUT(result, "output", STRING_OBJ(exec_impl(channel_id, src, opts, err)));
```

Comment:
Try this, I think this is why you're getting the memory leak. I believe you shouldn't return any allocated data when you're setting an error? And the allocation comes from `PUT`, which copies the `"output"` string.
```suggestion
  String output = exec_impl(channel_id, src, opts, err);
  if (ERROR_SET(err)) {
    return result;
  }
  PUT(result, "output", STRING_OBJ(output));
```

Reply:
Force-pushed this. All tests pass similarly locally: one `FAILED   test/functional/core/log_spec.lua @ 30: log messages are formatted with name or test id`.

@zeertzjq, could you, please, make the memory leak check again? Or how do I do it locally without bothering you?

Reply:
> CI checks for leaks, in the ASAN build. Looks like there aren't any now. You can build with ASAN enabled with `-DCLANG_ASAN_UBSAN=ON`

And now I know this too :)

Thanks for fixing the leak!